From 9819153bba6113594b342076ad2f323f173c3c28 Mon Sep 17 00:00:00 2001
From: NelloKudo <marshnelloosu@gmail.com>
Date: Wed, 6 Aug 2025 21:13:00 +0200
Subject: [PATCH] HACK: ws2_32: Block connection on first request with
 WINE_DISCONNECT_HACK=1

Zenless will do by default unless WINE_DISCONNECT_HACK=0 is set.
---
 dlls/ws2_32/socket.c | 15 +++++++++++++++
 1 file changed, 15 insertions(+)

diff --git a/dlls/ws2_32/socket.c b/dlls/ws2_32/socket.c
index d4223e93bd1..d9a6d53bbc7 100644
--- a/dlls/ws2_32/socket.c
+++ b/dlls/ws2_32/socket.c
@@ -1272,6 +1272,21 @@ int WINAPI connect( SOCKET s, const struct sockaddr *addr, int len )
     IO_STATUS_BLOCK io;
     HANDLE sync_event;
     NTSTATUS status;
+    
+    // HACK: Block first internet request with WINE_DISCONNECT_HACK
+    static int start = 0;
+    if (!start)
+    {
+        const char* env = getenv("WINE_DISCONNECT_HACK");
+        if (env)
+        {
+            start = 1;
+            FIXME("Disconnect env. detected, blocking first connection..\n");
+            return SOCKET_ERROR;
+        }
+
+        start = 1;
+    }
 
     TRACE( "socket %#Ix, addr %s, len %d\n", s, debugstr_sockaddr(addr), len );
 
-- 
2.50.1

